### PR TITLE
Sparse index: integrate with the `sparse-checkout` builtin

### PR DESCRIPTION
**Pull request by derrickstolee**
This integrates the `sparse-checkout` builtin with the sparse index. The tricky part here is that we need to partially expand the index when we are modifying the sparse-checkout definition.

Note that we modify the pattern list in a careful way: we create a `struct pattern_list` in-memory in `builtin/sparse-checkout.c` then apply those patterns to the index before writing the patterns to the sparse-checkout file. The `update_sparsity()` method does the work to assign the `SKIP_WORKTREE` bit appropriately, but this doesn't work if the files that are within the new sparse-checkout cone are still hidden behind a sparse directory.

The new `expand_to_pattern_list()` method does the hard work of expanding the sparse directories that are now within the new patterns. This expands only as far as needed, possibly creating new sparse directory entries.

This method does not contract existing files to sparse directories, and a big reason why is because of the check for ignored files as we delete those directories. The `clean_tracked_sparse_directories()` method is called after `update_sparsity()`, but we need to read the `A/B/.gitignore` file (or lack thereof) before we can delete `A/B/`. If we convert to sparse too quickly, then we lose this information and cause a full expansion.

Most of the correctness is handled by existing tests in `t1092`, but I add checks for `ensure_not_expanded` in some hopefully interesting cases.

As for performance, `git sparse-checkout set` can be slow if it needs to move a lot of files. However, no-op `git sparse-checkout set` (i.e. set the sparse-checkout cone to only include files at root, and do this on repeat) has these performance results on Linux in a monorepo with 2+ million files at `HEAD`:

```
Benchmark #1: baseline
  Time (mean ± σ):     10.465 s ±  0.018 s    [User: 9.885 s, System: 0.573 s]
  Range (min … max):   10.450 s … 10.497 s    5 runs
 
Benchmark #2: new code
  Time (mean ± σ):      68.9 ms ±   2.9 ms    [User: 45.8 ms, System: 17.1 ms]
  Range (min … max):    63.4 ms …  74.0 ms    41 runs
 
Summary
  'new code' ran
  151.89 ± 6.30 times faster than 'baseline'
```

### CODE REVIEW

File: sparse-index.c [hunk 1]
```diff
@@ -222,47 +227,115 @@ static int add_path_to_index(const struct object_id *oid,
...
 	/* initialize basics of new index */
 	full = xcalloc(1, sizeof(struct index_state));
 	memcpy(full, istate, sizeof(struct index_state));
 
+	/*
+	 * This slightly-misnamed 'full' index might still be sparse if we
+	 * are only modifying the list of sparse directories. This hinges
+	 * on whether we have a non-NULL pattern list.
+	 */
+	full->sparse_index = pl ? PARTIALLY_SPARSE : 0;
```
Comment (vdye) [inline sparse-index.c]:
```suggestion
	full->sparse_index = pl ? PARTIALLY_SPARSE : COMPLETELY_FULL;
```

I'm not sure whether the mix of enum & int is intentional here (if it is, add that to the "git style conventions I need to learn" list). If it isn't, though, I think spelling out the `COMPLETELY_FULL` value would help with readability.

File: sparse-index.c [hunk 2]
```diff
@@ -297,7 +378,7 @@ void ensure_full_index(struct index_state *istate)
 	/* Copy back into original index. */
 	memcpy(&istate->name_hash, &full->name_hash, sizeof(full->name_hash));
 	memcpy(&istate->dir_hash, &full->dir_hash, sizeof(full->dir_hash));
-	istate->sparse_index = 0;
+	istate->sparse_index = pl ? PARTIALLY_SPARSE : 0;
```
Comment (vdye) [inline sparse-index.c]:
```suggestion
	istate->sparse_index = pl ? PARTIALLY_SPARSE : COMPLETELY_FULL;
```

Same as the other comment - if what's currently there is the right way to write this, you can ignore/resolve this suggestion.

File: cache.h [hunk 3]
```diff
@@ -311,6 +311,28 @@ struct untracked_cache;
...
+	 * COMPLETELY_FULL: there are no sparse directories
+	 * in the index at all.
+	 */
+	COMPLETELY_FULL = 0,
+
+	/*
+	 * COMPLETELY_SPARSE: the sparse directories that exist
+	 * exactly match the sparse-checkout boundary.
+	 */
+	COMPLETELY_SPARSE = 1,
```
Comment (vdye) [inline cache.h]: I think there's a distinction between "completely collapsed" (as in, files outside the sparse checkout boundary might be in the index, but it's because they have modifications/are otherwise uncollapsible) and "completely sparse" (_nothing_ exists outside of the sparse checkout boundary). As far as I can tell, this setting refers to a state of "completely collapsed" (especially because `PARTIALLY_SPARSE` is described with "it is possible that some file entries could collapse to sparse directory entries."). The name `COMPLETELY_SPARSE` is probably fine as-is, but if it is really "completely collapsed", can you clarify that in the description?

File: sparse-index.c [hunk 4]
```diff
@@ -128,7 +133,7 @@ int convert_to_sparse(struct index_state *istate, int flags)
 {
 	int test_env;
 
-	if (istate->sparse_index || !istate->cache_nr ||
+	if (istate->sparse_index == COMPLETELY_SPARSE || !istate->cache_nr ||
```
Comment (vdye) [inline sparse-index.c]: This isn't quite the right place for me to comment (the relevant line is `read-cache.c:1831`), but it still somewhat relates.

There seem to be situations where the index can be in a partially sparse (in this case meaning "partially collapsed") state when it's read in, but because it's always read in as `COMPLETELY_SPARSE` or `COMPLETELY_FULL`, checks like this don't end up collapsing it. For example, using the setup from `t1092`:

```
init_repos &&
git checkout -b test-branch update-folder1 &&
git reset --mixed HEAD~1 &&
echo a >folder1/a
```

At the end of this block, the index can be further collapsed (removing `folder1`, since it now contains no staged or unstaged changes), but the index will be read into the next command as `COMPLETELY_SPARSE`. 

TL;DR not as much for this PR as for sparse index in general, but is it valuable to make the initial read of the index determine whether it's `PARTIALLY_SPARSE` or `COMPLETELY_SPARSE` (or just always read as `PARTIALLY_SPARSE`)? Would that cause too much of a performance hit? If not, then my suggestion would be to add a note to the description of `PARTIALLY_SPARSE` that it's an internal-only state.

**Review by vdye — APPROVED** [review]
My comments pretty much amount to documentation change requests, but I did have some questions about the implications of the updated `sparse_index_mode` on other commands.

File: sparse-index.c (hunk 4 above)
Comment (derrickstolee) [reply sparse-index.c]: You present a very interesting point: when we get these indexes that are not as collapsed as they could be, how do we get out of that state?

One way to do it would be to always assume `PARTIALLY_SPARSE` upon reading a new index. That currently leads to a performance drawback due to the recalculation of the cache-tree extension (perhaps we can avoid this?). I'll make a note to fix this later, as it is a problem that existed before the `PARTIALLY_SPARSE` mode.

File: cache.h (hunk 3 above)
Comment (derrickstolee) [reply cache.h]: Names are hard. I like your suggestion.
